[EDU-3829] feat: add manage objects in buckets guide (#626)
aziontech/docs: src/content/docs/en/pages/guides/object-storage/bucket-actions/index.mdx

---
title: How to upload and download objects from an Azion Object Storage bucket
description: Learn how to upload and download files from your object storage.
meta_tags: object storage, storage, cloud, s3, bucket, file storage
namespace: documentation_products_object_storage_upload

> permalink: /documentation/products/guides/how-to-upload-and-download-objects-from-bucket/

Reviewer: ```suggestion
permalink: /documentation/products/guides/upload-and-download-objects-from-bucket/
```

---

import Button from '~/components/Button.astro'

> **Azion Object Storage** allows you you to create buckets and integrate your object storage with your edge infrastructure. This guide covers how to upload and download objects from a bucket using the [Azion API](https://api.azion.com) and [Edge Runtime](/en/documentation/devtools/runtime/overview/).

Reviewer 2: ```suggestion
**Azion Object Storage** allows you to create buckets and integrate your object storage with your edge infrastructure. This guide covers how to upload and download objects from a bucket using the [Azion API](https://api.azion.com) and [Edge Runtime](/en/documentation/devtools/runtime/overview/).
```

> **Object Storage** allows you you to create buckets and integrate your object storage with your edge infrastructure. This guide covers how to upload and download objects from a bucket using the [Azion API](https://api.azion.com) and [Edge Runtime](/en/documentation/devtools/runtime/overview/).

Reviewer: ```suggestion
**Azion Object Storage** allows you you to create buckets and integrate your object storage with your edge infrastructure. This guide covers how to upload and download objects from a bucket using the [Azion API](https://api.azion.com) and [Edge Runtime](/en/documentation/devtools/runtime/overview/).
```

<Button href="/en/documentation/products/store/object-storage/" text="go to Object Storage reference documentation" variant="secondary"></Button>

> Read [How to create and modify an Azion Object Storage bucket](/en/documentation/products/guides/how-to-upload-and-download-objects-from-bucket/) for bucket operations.

Reviewer: não seria o link do outro guia?

> Read [How to create and modify an Azion Object Storage bucket](/en/documentation/products/guides/how-to-upload-and-download-objects-from-bucket/) for bucket operations.

Author: ```suggestion
Read [How to create and modify an Azion Object Storage bucket](/en/documentation/products/guides/create-and-modify-bucket/) for bucket operations.
```

---

## Via API

### Upload an object to a bucket

To upload an object, run the following `POST` request in your terminal, replacing `[TOKEN VALUE]` with your [personal token](/en/documentation/products/guides/personal-tokens/), `<bucket_name>` with the name of your bucket, `<object_key>` with an ID or name for the object, the `Content-Type` header with the MIME type being submitted, and the object sent as a data binary:

```bash
curl --location 'https://api.azion.com/v4/storage/buckets/<bucket_name>/objects/<object_key>' \
--header 'Accept: application/json; version=3' \
--header 'Content-Type: text/csv' \
--header 'Authorization: Token [TOKEN VALUE]' \
--data '@./path/file.csv'
```

You should receive the following response:

```json
{
    "state": "executed",
    "data": {
        "object_key": "folder/csv-file"
    }
}
```

> ### Download an object from a bucket

Reviewer: ```suggestion
### Downloading an object from a bucket
```

To download the object, run the following `GET` request in your terminal, replacing `[TOKEN VALUE]` with your [personal token](/en/documentation/products/guides/personal-tokens/), `<bucket_name>` with the name of your bucket, and `<object_key>` with the key created for the object:

```bash
curl --location 'https://api.azion.com/v4/storage/buckets/<bucket_name>/objects/<object_key>' \
--header 'Accept: application/json; version=3' \
--header 'Authorization: Token [TOKEN VALUE]'
```

---

## Via Edge Runtime

You can create an edge function to upload objects to your bucket using a `POST` console command and see if an object is available using a `GET` console command. To do so:

> 1. Access [Real-Time Manager (RTM)](https://manager.azion.com/) > **Edge Functions**.

Reviewer 2: ```suggestion
1. Access [Real-Time Manager (RTM)](/en/documentation/products/guides/how-to-access-rtm/) > **Edge Functions**.
```

2. Click the **Add Function** button.

> 3. Name your function. Example: `my-bucket POST+GET`

Reviewer: ```suggestion
3. Name your function. Example: `my-bucket POST+GET`.
```

> 4. In the **Code** tab, add the following Javascript code: 

Reviewer: ```suggestion
4. In the **Code** tab, add the following JavaScript code:
```

    return new Response(await asset.arrayBuffer(), {
        headers: {
            "Content-Type": asset.contentType
        },
    });
}

async function doPost(path, content_type, value, bucket_name) {
    console.log(`Path: ${path}`);
    let options = {
        "Content-Type": content_type
    }
    console.log(`Options: ${options}`);
    const storage = new Storage(bucket_name);
    await storage.put(path, value, options);
    return new Response("Object added.");
}

async function router(event) {
    const request = event.request;
    const method = request.method;
    const path = decodeURI(new URL(request.url).pathname);
    const bucket_name = event.args.bucket;
    if (method === "POST") {
        let content_type = request.headers.get("Content-Type");
        let content = await request.arrayBuffer();
        return doPost(path, content_type, content, bucket_name);
    } else if (method === "GET") {
        return doGet(path, bucket_name);
    } else {
        throw new Error(`Invalid method: ${method}. Expected POST or GET.`);
    }
}
addEventListener("fetch", (event) => {
    event.respondWith(
        router(event)
    );
});
```

> | Variable | Expected value |

Reviewer: o label ficou Expect value mas tem description + examples

> | Variable | Expected value |

Author: ```suggestion
| Variable | Description |
```

| --- | --- |
| `path` | Path to the object. Example: `./path/file.csv` |
| `bucket_name` | The name of the bucket. Example: `my-bucket` |
| `content_type` | The MIME type of the object. Example: `text/csv` |
| `value` | The contents of the object as a data binary |

5. In the **Arguments** tab, add the object with the `bucket` property and, as a value, the name of the bucket as a string.

```json
{
  "bucket": "my-bucket"
}
```

6. Click the **Save** button.

:::note
See the [Object Storage API](/en/documentation/devtools/runtime/api-reference/storage/) reference for more details on functions available for other types of requests.
:::

> Once you have the edge function ready, you need to create an edge application that will proxy the upload process for the Object Storage.

Reviewer: instância do object storage, em minúsculo?

> Once you have the edge function ready, you need to create an edge application that will proxy the upload process for the Object Storage.

Author: ```suggestion
Once you have the edge function ready, you need to create an edge application that will proxy the upload process for the bucket.
```

> 1. Access [Real-Time Manager (RTM)](https://manager.azion.com/) 

Reviewer: ```suggestion
1. Access [Real-Time Manager (RTM)](https://manager.azion.com/).
```

2. Select the **Start with a template** option on the homepage.

> 3. Select the option **Build from scratch**.

Reviewer 2: ```suggestion
3. Select the **Build from scratch** option.
```
